tidb/src/main/java/io/tidb/bigdata/tidb/TiDBEncodeHelper.java

```diff
@@ -0,0 +1,337 @@
+    byte[] value;
+    if (remove) {
+      value = EMPTY_BYTES;
+    } else {
+      value = ZERO_BYTES;
+    }
+    return new BytePairWrapper(codecDataOutput.toBytes(), value);
+  }
+
+  private BytePairWrapper generateRecordKeyValue(Row tiRow, long handle, boolean remove) {
+    byte[] key;
+    byte[] value;
+    if (remove) {
+      key = RowKey.toRowKey(tiTableInfo.getId(), handle).getBytes();
+      value = new byte[0];
+    } else {
+      key = RowKey.toRowKey(tiTableInfo.getId(), handle).getBytes();
+      value = encodeTiRow(tiRow);
+    }
+    return new BytePairWrapper(key, value);
+  }
+
+  private List<BytePairWrapper> generateIndexKeyValues(Row tiRow, long handle,
+      boolean remove) {
+    return tiTableInfo.getIndices().stream()
+        .filter(tiIndexInfo -> !(isCommonHandle && tiIndexInfo.isPrimary()))
+        .map(tiIndexInfo -> {
+          if (tiIndexInfo.isUnique()) {
+            return generateUniqueIndexKeyValue(tiRow, handle, tiIndexInfo, remove);
+          } else {
+            return generateSecondaryIndexKeyValue(tiRow, handle, tiIndexInfo, remove);
+          }
+        }).collect(Collectors.toList());
+  }
+
+  private List<BytePairWrapper> generateDataToBeRemoved(Row tiRow, long handle) {
+    Snapshot snapshot = session.getTiSession().createSnapshot(timestamp.getPrevious());
+    List<Pair<Row, Long>> deletion = new ArrayList<>();
+    if (handleCol != null || isCommonHandle) {
+      byte[] key = RowKey.toRowKey(tiTableInfo.getId(), handle).getBytes();
+      byte[] oldValue = snapshot.get(key);
+      if (!isEmptyArray(oldValue) && !isNullUniqueIndexValue(oldValue)) {
+        Row oldRow = TableCodec.decodeRow(oldValue, handle, tiTableInfo);
+        deletion.add(new Pair<>(oldRow, handle));
+      }
+    }
+    for (TiIndexInfo index : uniqueIndices) {
+      if (!isCommonHandle || !index.isPrimary()) {
+        Pair<byte[], Boolean> uniqueIndexKeyPair = buildUniqueIndexKey(tiRow, handle, index);
+        if (!uniqueIndexKeyPair.second) {
+          byte[] oldValue = snapshot.get(uniqueIndexKeyPair.first);
+          if (!isEmptyArray(oldValue) && !isNullUniqueIndexValue(oldValue)) {
+            long oldHandle = TableCodec.decodeHandle(oldValue);
+            byte[] oldRowValue = snapshot.get(
+                RowKey.toRowKey(tiTableInfo.getId(), oldHandle).getBytes());
+            Row oldRow = TableCodec.decodeRow(oldRowValue, oldHandle, tiTableInfo);
+            deletion.add(new Pair<>(oldRow, oldHandle));
+          }
+        }
+      }
+    }
+    List<BytePairWrapper> deletionKeyValue = new ArrayList<>();
+    for (Pair<Row, Long> pair : deletion) {
+      deletionKeyValue.add(generateRecordKeyValue(pair.first, pair.second, true));
+      deletionKeyValue.addAll(generateIndexKeyValues(pair.first, pair.second, true));
+    }
+    return deletionKeyValue;
+  }
+
+  public List<BytePairWrapper> generateKeyValuesByRow(Row row) {
+    Preconditions.checkArgument(row.fieldCount() == columns.size(),
```

Comment:
It seems that we can't omit auto random column. is this reasonable

Reply:
@humengyu2012

Reply:
There is a config named `tikv.sink.ignore-autoincrement-column-value`. If true, we will ignore all raw values of auto increment column and auto generate it; if false we will use the raw value. However, we do not support the coexistence of null and non null values for `tikv.sink.ignore-autoincrement-column-value=false`.